py/src/braintrust/util.py

```diff
@@ -50,3 +52,31 @@ def get_caller_location():
 # https://stackoverflow.com/questions/5574702/how-do-i-print-to-stderr-in-python.
 def eprint(*args, **kwargs):
     print(*args, file=sys.stderr, **kwargs)
+
+
+T = TypeVar("T")
+
+
+class LazyValue(Generic[T]):
+    """A simple wrapper around a callable object which computes the value
+    on-demand and saves it for future retrievals.
+    """
+
+    def __init__(self, callable: Callable[[], T], use_mutex: bool):
+        self.callable = callable
+        self.mutex = threading.RLock() if use_mutex else None
```

Comment:
is there potentially some benefit to using an `RWLock` here? Also why does it need to be re-entrant? In case a lazy value depends on itself?

Reply:
Hmm, you're right that re-entrancy is unnecessary overhead here. I'm not sure RWLock is the right solution to reduce overhead (would we acquire a read lock and then "upgrade" it if the value hasn't been computed yet?)

I'm thinking the right thing to do is a short-circuit check, like
```
if self.has_computed:
  return self.value
with self.lock:
  if not self.has_computed:
    ...
return self.value
```

Pretty sure we're not going to run into sequential consistency issues in python (i.e if I read `self.has_computed == True` I'm guaranteed to pick up the new value of `self.value`)? Couldn't find docs but I think the GIL should ensure that?

Reply:
Yes it will ensure sequential consistency. I think that approach, with a plain mutex (or even re-entrant), is fine, since now the average overhead is much much lower.